src/nupic/ntypes/Sdr.hpp

```diff
@@ -754,26 +766,99 @@ class SparseDistributedRepresentation : public Serializable
         // Read the dimensions.
         readVector( dimensions_ );
 
-        // Read the data.
-        clear();
-        readVector( flatSparse );
-        flatSparse_valid = true;
-
-        // Consume the end marker.
-        inStream >> marker;
-        NTA_ASSERT( marker == "~SDR" );
-
         // Initialize the SDR.
         // Calculate the SDR's size.
         size_ = 1;
         for(UInt dim : dimensions)
             size_ *= dim;
         // Initialize sparse tuple.
         sparse.assign( dimensions.size(), {} );
+
+        // Read the data.
+        readVector( flatSparse );
+        setFlatSparseInplace();
+
+        // Consume the end marker.
+        inStream >> marker;
+        NTA_ASSERT( marker == "~SDR" );
     };
+
+    void addCallback(SDR_callback_t callback) {
+        callbacks.push_back( callback );
+    }
 };
 
 typedef SparseDistributedRepresentation SDR;
 
+class SDR_Proxy : public SDR
+{
+public:
+    SDR_Proxy(SDR &sdr, const vector<UInt> &dimensions)
+        : SDR( dimensions ) {
+        parent = &sdr;
+        NTA_CHECK( size == parent->size ) << "SDR Proxy must have same size as given SDR.";
+        clear();
+        parent->addCallback( (SDR_callback_t) this );
+    };
+
+    SDR_dense_t& getDense()
+        { return parent->getDense(); }
+
+    SDR_flatSparse_t& getFlatSparse()
+        { return parent->getFlatSparse(); }
+
+    void zero()
+        { NTA_THROW << _SDR_Proxy_setter_error_message; };
+
+    void setDense( SDR_dense_t &value )
```

Comment:
can you hide these methods instead? Ie by making them `private`.

Reply:
best would be private and the existing throw.

Reply:
Instead of making them private I overloaded the set*Inplace methods to throw an exception, since all of the setters call one of the set*Inplace methods.  This makes the code much shorter.  It also compiles, c++ didn't like virtual templates for some reason.